Sources/MongoSwift/MongoCollection+FindAndModify.swift

```diff
@@ -50,32 +56,61 @@ extension MongoCollection {
      *
      * - Returns: A `CollectionType` representing either the original or updated document,
      *      depending on selected options, or `nil` if there was no match.
-     * - Throws: 
+     *
+     * - Throws:
+     *   - `MongoError.invalidArgument` if any of the provided options are invalid
      *   - `MongoError.commandError` if there are any errors executing the command.
      *   - A `DecodingError` if the updated document cannot be decoded to a `CollectionType` value
      */
     @discardableResult
     public func findOneAndUpdate(filter: Document, update: Document,
                                  options: FindOneAndUpdateOptions? = nil) throws -> CollectionType? {
-        throw MongoError.commandError(message: "Unimplemented command")
+        return try self.findAndModify(filter: filter, update: update, options: options)
+    }
+
+    /// A private helper method for findAndModify operations to use
+    private func findAndModify(filter: Document, update: Document? = nil,
+                               options: FindAndModifyOptionsConvertible? = nil) throws -> CollectionType? {
+
+        var opts = try options?.asOpts()
+
+        if let update = update {
+            // if no options were provided but we need to set update, create them
+            opts = opts ?? FindAndModifyOptions()
+            try opts!.setUpdate(update)
+        }
+
+        let reply = Document()
+        var error = bson_error_t()
+
+        if !mongoc_collection_find_and_modify_with_opts(self._collection,
+        filter.data, opts?._options, reply.data, &error) {
+            // TODO SWIFT-144: replace with more descriptive error type(s)
+            throw MongoError.commandError(message: toErrorString(error))
+        }
+
+        guard let value = reply["value"] as? Document else { return nil }
```

Comment:
Just realized that `mongoc_collection_find_and_modify_with_opts()` doesn't unwrap the `value` field for us. Oh well.